### PR TITLE
Added optional flag for building Arrow with S3 filesystem support

### PR DESCRIPTION
**Pull request by jdye64**
Added the cmake build flag `CUDF_ENABLE_ARROW_S3` to conditionally build Arrow with S3 filesystem support. By default support for S3 filesystem will be enabled. If S3 support needs to be disabled you can pass `-DCUDF_ENABLE_ARROW_S3=OFF`

### CODE REVIEW

**Review by jlowe — APPROVED** [review]
Java approval

File: cpp/cmake/thirdparty/CUDF_GetArrow.cmake [hunk 1]
```diff
@@ -14,11 +14,12 @@
 # limitations under the License.
 #=============================================================================
 
-function(find_and_configure_arrow VERSION BUILD_STATIC)
+function(find_and_configure_arrow VERSION BUILD_STATIC ENABLE_S3)
 
     set(ARROW_BUILD_SHARED ON)
     set(ARROW_BUILD_STATIC OFF)
     set(CPMAddOrFindPackage CPMFindPackage)
+    set(ARROW_BUILD_S3 OFF)
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/CUDF_GetArrow.cmake:22]: Nitpick, but can you move this a line above `CPMAddOrFindPackage` so it is with the other `ARROW_BUILD` variables.

**Review by robertmaynard — APPROVED** [review]

**Review by robertmaynard — COMMENTED** [review]
What happens when S3 is disabled and `arrow_io_source` is called with an S3 backed URI? Should we have a test that we get the expected exception and don't crash?

**Comment by jdye64** [comment]
> What happens when S3 is disabled and `arrow_io_source` is called with an S3 backed URI? Should we have a test that we get the expected exception and don't crash?

In that case https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/io/datasource.hpp#L321 would trigger. Always good to add a test however.

**Review by mythrocks — APPROVED** [review]

**Comment by trxcllnt** [comment]
@gpucibot merge

**Pull request merged**